creator/data_templates/models.py

```diff
@@ -0,0 +1,181 @@
+
+User = get_user_model()
+
+
+def template_id():
+    return kf_id_generator("DT")
+
+
+def template_version_id():
+    return kf_id_generator("TV")
+
+
+class DataTemplate(models.Model):
+    """
+    A data submission template with versions
+
+    The Template model represents the essence of the data template just like
+    File does with Version, and it provides a container for the versions
+    of the template
+    """
+
+    class Meta:
+        permissions = [
+            ("list_all_datatemplate", "Show all data_template"),
+        ]
+
+    id = models.CharField(
+        max_length=11,
+        primary_key=True,
+        default=template_id,
+        help_text="Human friendly ID assigned to the data template",
+    )
+    uuid = models.UUIDField(
+        default=uuid.uuid4, help_text="UUID used internally"
+    )
+    created_at = models.DateTimeField(
+        auto_now_add=True,
+        null=True,
+        help_text="Time when the data template was created",
+    )
+    modified_at = models.DateTimeField(
+        auto_now=True,
+        null=True,
+        help_text="Time when the data template was modified",
+    )
+    organization = models.ForeignKey(
+        Organization,
+        null=False,
+        blank=False,
+        on_delete=models.CASCADE,
+        related_name="data_templates",
+        help_text="The organization that owns the data template",
+    )
+    creator = models.ForeignKey(
+        User,
+        null=True,
+        blank=True,
+        on_delete=models.SET_NULL,
+        related_name="data_templates",
+        help_text="The user who created the data template",
+    )
+    name = models.CharField(
+        max_length=256, help_text="Name of the data template"
+    )
+    description = models.TextField(
+        max_length=10000,
+        help_text="Description of data template",
+    )
+    icon = models.CharField(
+        max_length=256,
+        null=True,
+        blank=True,
+        help_text="Name of the Font Awesome icon to use when displaying the "
+        "template in the frontend web application",
+    )
+
+    @property
+    def released(self):
+        """
+        Determines whether the data template is being used in any studies
+
+        A data template is being used in a study if one or more of its
+        versions has been assigned to a study
+        """
+        assigned_to_studies = (
+            TemplateVersion.objects.filter(data_template__pk=self.pk)
+            .exclude(studies=None)
+            .count()
+        )
+        return assigned_to_studies > 0
+
+
+class TemplateVersion(models.Model):
+    """
+    A version of a specific data submission template
+
+    Encapsulates the data template's field definitions (e.g. columns,
+    accepted values, instructions for populating, etc.)
+    """
+
+    class Meta:
+        permissions = [
+            ("list_all_templateversion", "Show all template_version"),
+        ]
+
+    id = models.CharField(
```

Comment:
where/how is this version created? Is it obvious that one version comes after another?

Reply:
I'm not sure what you mean by the first question..you mean like what mutation or how will the user create the version from the UI?

From this ID, it won't be obvious that one version comes after another. I think we might want to have a separate semantic version number on the `TemplateVersion` for that. I didn't add that yet bc there are complexities in determining the rules for how to assign the semantic version number for a new version.

Reply:
oh I meant field. How is this field set? I don't see it in the Input objects.

Reply:
Ohh I see. Its automatically generated by the function passed to the default kwarg, `default=template_version_id`. Is that what you mean?

Reply:
oh yea I didn't get what that meant - but I see the functions above now